### PR TITLE
175-Add a pattern option to the identifier data type adapter

### LINKED ISSUE #175: Add a `pattern` option to the `identifier` data type adapter
Some identifiers have mandated display formats - e.g.:
 - US SSN: 999-99-999 (https://en.wikipedia.org/wiki/Social_Security_number)
 - US MBI: https://noworldborders.com/2020/01/08/medicare-beneficiary-identifier-required-in-2020/
 - CA SIN: https://en.wikipedia.org/wiki/Social_insurance_number

The `identifier` datatype adapter could take an optional `pattern` argument that can provide a pattern to format the values.
The pattern could match numbers and/or letters to present them in groups with separators. (format TBD).

The pattern could also accept a mapping of system and patterns, so that applications could have a constant somewhere that is passed systematically to the formatter everywhere, and the appropriate pattern is applied based on the identifier system.

A default pattern value could be included that have a number of patterns defined OOB for common systems.

Example usage:

```typescript
const identifier: Identifier = {
  system: "http://hl7.org/fhir/sid/us-ssn",
  value: "75359831"
};

dataTypeAdapter.identifier.format(identifier, { pattern: "999-99-999" }) -> 753-59-831

// Default OOB patterns
dataTypeAdapter.identifier.format(identifier) -> 753-59-831

// Custom app mapping
const IdentifierPatterns: Record<string, string> = {
  "http://hl7.org/fhir/sid/us-ssn": "999-99999",
  "http://my-custom-identifier-system": "999-AA-BBB",
};

dataTypeAdapter.identifier.format(identifier, { pattern: IdentifierPatterns }) -> 753-59831

```

### PR DESCRIPTION
**Pull request by jadiaheno**
## What is this about

A `pattern` control was added to the options object of the `fhirIdentifierTypeAdapter.format()` function. This pattern control was added to support mandatory display formats of FHIR identifiers. 

The supported values are:
- a string: `999-9*9-aaaa`
- a map: `{ "https://fhir.nhs.uk/Id/nhs-number": "999 999 9999" }`
- a boolean: `true` for default formatting

The default format characters are:
- `9` for numbers
- `a` for letters
- `*` for numbers and letters
Any other character will be used as a template.

The formatting function can be reused and is in [utils](https://github.com/bonfhir/bonfhir/pull/199/files#diff-b72391cd4891945ceafd5e7837e4f9b691b2389b29d966ee18746fc064b4b793).
image

## Issue ticket numbers

#175 

## How can this be tested?

`yarn run test`

## Known limitations/edge cases

As of right now, the formatting is `at best`. It provides flexibility for custom patterning. Some limitations can be introduced  if we want to reject patterns (returning the original value)

image

### CODE REVIEW

File: packages/core/r4b/data-types/identifier.test.ts [hunk 1]
```diff
@@ -170,4 +170,98 @@ describe("fhirIdentifierTypeAdapter", () => {
...
+        value: "75359831",
+      };
+      expect(
+        fhirIdentifierTypeAdapter().format(identifier, {
+          pattern: "",
+          style: "value",
+        })
+      ).toEqual("75359831");
+    });
+  });
```
Comment (julienblin) [inline packages/core/r4b/data-types/identifier.test.ts:312]: This test does not pass at the moment:

```typescript
it("does nothing if already formatted", () => {
      const identifier: Identifier = {
        system: "http://hl7.org/fhir/sid/us-ssn",
        value: "753-59-831",
      };
      expect(
        fhirIdentifierTypeAdapter().format(identifier, {
          pattern: true,
          style: "value",
        })
      ).toEqual("753-59-831");
    });
```

File: packages/core/r4b/data-types/identifier.ts [hunk 2]
```diff
@@ -49,6 +50,25 @@ export type FhirIdentifierFormatOptions = {
...
+   * @example
+   * An us-mbi identifier will be formatted with the default system pattern
+   * // returns "1EG4-TE5-MK73"
+   * fhirIdentifierTypeAdapter.format(identifier, { pattern: true })
+   * @example
+   * Identifier with "https://fhir.nhs.uk/Id/nhs-number"  as system using the provided system:pattern mapping
+   * // returns "123 456 7890"
+   * fhirIdentifierTypeAdapter.format(identifier, { pattern: { "https://fhir.nhs.uk/Id/nhs-number": "999 999 9999" } })
+   */
+  pattern?: string | Record<string, string> | boolean | null | undefined;
```
Comment (julienblin) [inline packages/core/r4b/data-types/identifier.ts]: By default it should format using the default patterns.
I would suggest only accepting `false` as a value to disable the default patterns.


```suggestion
  pattern?: string | Record<string, string> | false | null | undefined;
```

File: packages/core/r4b/utils.ts [hunk 3]
```diff
@@ -61,3 +61,43 @@ export function asArray<T>(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   return Array.isArray(value) ? (value as any) : [value];
 }
+
+/**
+ * Default characters used for pattern value formatting.
+ */
+const defaultFormatChars: { [key: string]: RegExp } = Object.freeze({
```
Comment (julienblin) [inline packages/core/r4b/utils.ts]: How can we escape does characters so that we can use it in the output (instead of a placeholder)?

File: packages/core/r4b/data-types/identifier.ts [hunk 4]
```diff
@@ -233,3 +266,13 @@ export const DEFAULT_SYSTEMS_LABELS: Required<
...
+
+/**
+ * Default values for `systemsPatterns`.
+ */
+export const DEFAULT_SYSTEMS_PATTERNS: Required<Record<string, string>> = {
+  "http://hl7.org/fhir/sid/us-ssn": "999-99-999",
+  "http://hl7.org/fhir/sid/us-mbi": "9a*9-a*9-aa99",
+  "https://fhir.infoway-inforoute.ca/NamingSystem/ca-social-insurance-number":
+    "999-999-999",
+};
```
Comment (julienblin) [inline packages/core/r4b/data-types/identifier.ts]: We should remove the Canadian SIN as it's not part of the official spec.
```suggestion
export const DEFAULT_SYSTEMS_PATTERNS: Required<Record<string, string>> = {
  "http://hl7.org/fhir/sid/us-ssn": "999-99-999",
  "http://hl7.org/fhir/sid/us-mbi": "9a*9-a*9-aa99",
};
```

File: packages/core/r4b/data-types/identifier.ts [hunk 5]
```diff
@@ -49,6 +50,25 @@ export type FhirIdentifierFormatOptions = {
...
+   * fhirIdentifierTypeAdapter.format(identifier, { pattern: true })
+   * @example
+   * Identifier with "https://fhir.nhs.uk/Id/nhs-number"  as system using the provided system:pattern mapping
+   * // returns "123 456 7890"
+   * fhirIdentifierTypeAdapter.format(identifier, { pattern: { "https://fhir.nhs.uk/Id/nhs-number": "999 999 9999" } })
+   */
+  pattern?: string | Record<string, string> | boolean | null | undefined;
 };
 
 export interface FhirIdentifierTypeAdapter {
```
Comment (julienblin) [inline packages/core/r4b/data-types/identifier.ts:80]: We should accept a global pattern map, the same way we do with `systemsLabels`, so that the default can be overriden globally on a specific adapter instance.

**Review by julienblin — CHANGES_REQUESTED** [review]
This is great. A couple of changes though:

- default patterns should apply by default; settings to false disable patterns if need be
- we should be able to set a different set of default patterns for a specific instance of an adapter, same as `systemsLabels` on the constructor
- we should handle properly the case of having the source value already formatted correctly (see missing test case)
- we should be able to escape pattern characters to use them in the output string if need be

Maybe we could simplify the pattern by only using `#` as a placeholder? We don't care about validating the values, just formatting them.

So what about using only `#` as a placeholder, and having the possibility to use `\#` to output `#` in the final pattern?

File: packages/core/r4b/data-types/identifier.ts (hunk 2 above)
Comment (jadiaheno) [reply packages/core/r4b/data-types/identifier.ts]: Default formatting using the default patterns would be a breaking change of the format function?

Comment (julienblin) [reply packages/core/r4b/data-types/identifier.ts]: Yes. In this case it's OK though, the impact is minimal and we are still in alpha :-)

File: packages/core/r4b/utils.ts (hunk 3 above)
Comment (jadiaheno) [reply packages/core/r4b/utils.ts]: Per recommendation, escaping with `\`.

File: packages/core/r4b/data-types/identifier.test.ts (hunk 1 above)
Comment (jadiaheno) [reply packages/core/r4b/data-types/identifier.test.ts:312]: Added.

File: packages/core/r4b/data-types/identifier.ts (hunk 4 above)
Comment (jadiaheno) [reply packages/core/r4b/data-types/identifier.ts]: CA SIN removed.

File: packages/core/r4b/data-types/identifier.ts (hunk 5 above)
Comment (jadiaheno) [reply packages/core/r4b/data-types/identifier.ts:80]: `systemsPatterns` added.

**Comment by jadiaheno** [comment]
I don't have any further changes at this time. Once the PR is ready to be merged, I'll make sure to rebase it accordingly.

**Review by julienblin — CHANGES_REQUESTED** [review]
This is great - thanks for the fixes!

I have one last request related to https://bonfhir.dev/contributing/contributing#conventional-commits

Can you please merge all this, rebase and create a single nice conventional commit?
The reason is we use this to generate changelogs, so a nice clean one is important.
The rest is great, exactly what we need - thank you!

**Comment by jadiaheno** [comment]
The changes were squashed into a single commit.

**Review by julienblin — APPROVED** [review]
LGTM 🎉
